presto-main-base/src/main/java/com/facebook/presto/sql/planner/optimizations/MergePartialAggregationsWithFilter.java

```diff
@@ -218,17 +228,32 @@ else if (node.getStep().equals(FINAL)) {
-
-            context.get().getPartialResultToMask().putAll(aggregationsToMergeOutput.entrySet().stream()
-                    .collect(toImmutableMap(x -> x.getValue(), x -> x.getKey().getMask().get())));
-            context.get().getPartialOutputMapping().putAll(aggregationsToMergeOutput.entrySet().stream()
-                    .collect(toImmutableMap(x -> x.getValue(), x -> aggregationsWithoutMaskToOutput.get(removeFilterAndMask(x.getKey())))));
+                    .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+
+            List<List<AggregationNode.Aggregation>> aggregationWithMaskNotMatched = node.getAggregations().entrySet().stream().map(Map.Entry::getValue).filter(x -> x.getMask().isPresent() && !aggregationsToMergeOutput.containsKey(x))
+                    .collect(Collectors.groupingBy(AggregationNodeUtils::removeFilterAndMask)).values()
+                    .stream().filter(x -> x.size() > 1).collect(Collectors.toList());
+            List<AggregationNode.Aggregation> newAggregationNoMask = aggregationWithMaskNotMatched.stream().map(x -> removeFilterAndMask(x.get(0))).collect(Collectors.toList());
+            List<VariableReferenceExpression> variableForNewAggregationNoMask = newAggregationNoMask.stream().map(x -> variableAllocator.newVariable(x.getCall())).collect(Collectors.toList());
+            Map<VariableReferenceExpression, AggregationNode.Aggregation> aggregationsAdded = IntStream.range(0, newAggregationNoMask.size()).boxed().collect(toImmutableMap(variableForNewAggregationNoMask::get, newAggregationNoMask::get));
+
+            Map<AggregationNode.Aggregation, VariableReferenceExpression> aggregationsWithoutMaskToOutputCombined = ImmutableMap.<AggregationNode.Aggregation, VariableReferenceExpression>builder().putAll(aggregationsWithoutMaskToOutput)
+                            .putAll(aggregationsAdded.entrySet().stream().collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey))).build();
+            Map<AggregationNode.Aggregation, VariableReferenceExpression> aggregationsToMergeOutputCombined = node.getAggregations().entrySet().stream()
+                    .filter(x -> x.getValue().getMask().isPresent() && aggregationsWithoutMaskToOutputCombined.containsKey(removeFilterAndMask(x.getValue())))
+                    .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+
+            context.get().getNewAggregationOutput().addAll(variableForNewAggregationNoMask);
+            context.get().getPartialResultToMask().putAll(aggregationsToMergeOutputCombined.entrySet().stream()
+                    .collect(toImmutableMap(Map.Entry::getValue, x -> x.getKey().getMask().get())));
+            context.get().getPartialOutputMapping().putAll(aggregationsToMergeOutputCombined.entrySet().stream()
+                    .collect(toImmutableMap(Map.Entry::getValue, x -> aggregationsWithoutMaskToOutputCombined.get(removeFilterAndMask(x.getKey())))));
```

Comment:
Can we actually add the OR of the filters in this case?

So SUM(IF(p1, x)), SUM(IF(p2, x)) 

will have

SUM(IF(p1 or p2, x)) 

in the partial agg.

Reply:
Done